### ISSUE TITLE
Server hang on block updates (getClaimAt)

### ISSUE THREAD

**Issue by MrArcane** [GriefPrevention/GriefPrevention#66]
### What behavior is expected:
My server not randomly crashing
### What behavior is observed:
The server randomly crashing
### Steps/models to reproduce:
I turned the server on 2 different times with the same output as before

### Spigot/Craftbukkit/Paper and GriefPrevention version:
Spigot 1.11.2
GriefPrevention 16.5

### Stack trace/error (if applicable)
http://pastebin.com/rKaaPAST

### GriefPrevention config.yml (if applicable)
http://pastebin.com/4KywKb02

### Plugin list (if applicable):
http://prntscr.com/dxh2j5

**Comment by sgtcaze** [comment]
Saw something similar, except my stacktrace is:

```
[21:42:58] [Spigot Watchdog Thread/ERROR]: 		me.ryanhamshire.GriefPrevention.DataStore.getClaimAt(DataStore.java:673)
[21:42:58] [Spigot Watchdog Thread/ERROR]: 		me.ryanhamshire.GriefPrevention.BlockEventHandler.onBlockFromTo(BlockEventHandler.java:714)
```

The common thread is that when a block movement/update occurs, the claim for the block is determined here https://github.com/TechFortress/GriefPrevention/blob/master/src/me/ryanhamshire/GriefPrevention/DataStore.java#L670. However, this hangs the server.

**Comment by RoboMWM** [comment]
And you both have been able to confirm this doesn't occur on 16.4?

Only culprit I could think of right now (if you're able to confirm the above) is somehow @BillyGalbreath's PR (#48) is causing issues. Likely something to do with that `Collections<Claim>` stuff.

**Comment by BillyGalbreath** [comment]
The stacktrace doesnt touch the Visualization code I added. The stacktrace is hanging on piston extract/retract trying to find the claims of two locations at once. Probably someone being naughty with piston a piston on a clock rather than plugin fault, but I cant seem to reproduce.

Either that or something with player chat is locking up on the new DeniedMessageEvent. Almost want to bet money its from AsyncChatEvent calling synchronized DataStore#getMessage()

@MrArcane can you show more lines before the crash from the logs? I'm sure someone said the words "stuck" or "trapped" approx 60 seconds before the crash.

http://paste.md-5.net/faxigeyona.avrasm

**Comment by sgtcaze** [comment]
I doubt it would be the new DeniedMessageEvent. I took a look at the synchronized DataStore#getMessage method and personally do not see how that would hang the server. However, I can make a test version and drop the synchronized keyword and see if this changes anything on our side.

I'd wager it's to do with the BlockEventHandler class checking claims. Even still, we haven't been able to reproduce.

**Comment by RoboMWM** [comment]
Also potentially related: https://www.spigotmc.org/threads/griefprevention.35615/page-176#post-2173672

I will need a full server log for any future reports on this issue (I have some ideas about how this is being caused, but will need to see the server log to confirm since I have not been able to reproduce such).

**Comment by MrArcane** [comment]
Well I posted it on my test server and tried a couple of the chat things and it didn't crash, I also tried running some things people said before it happened, But here is the full log.

http://pastebin.com/FESnam8f

**Comment by Dreadonator** [comment]
@RoboMWM 

Today my server also crashed after updating. 
Maybe it is related. Full log is here:
https://github.com/TechFortress/GriefPrevention/issues/68

**Comment by RoboMWM** [comment]
@Dreadonator will need full log pasted in a paste service such as http://gist.github.com

@ everyone else: Neither Billy nor I have been able to reproduce this issue. Please try to reproduce this issue, and post the steps required to do so.

**Comment by plasticblues** [comment]
Just had this happen to me, trying to figure out who did it and what they did to crash but heres my log for the moment:
http://pastebin.com/1JQ4SH7Z

Server crashed again, person was using an enderman farm that uses pistons to push mobs down it via pressure plates! The claim is all 1 claim. Trying to figure out if anyone else was using pistons but that seems to be the consistent thing, the server crashed, person logged in, was at enderman farm, server crashed again.

**Comment by RoboMWM** [comment]
Cool @plasticblues can you reproduce on a test server? (I currently cannot at this moment)

**Comment by plasticblues** [comment]
Not at home at the moment, I can test after work in about 4 hours (if nothing goes wrong at work) if it hasn't been fixed.

**Comment by electronicboy** [comment]
The issue looks like it specifically boils down to the sync lock on the method, @BillyGalbreath's comment of the chat threads lead me onto something I was looking into from md's comment on IRC, 

https://github.com/TechFortress/GriefPrevention/commit/0d26d6c65bf38de567cd4eca75f7feba136b0417 appears to be the commit that broke stuff, Specifically the calling of the event, I have a feeling that there is a tangle in the synchronization in the event system, combined with the synchronization in the DataStore class

**Comment by RoboMWM** [comment]
Potentially, though @BillyGalbreath tested that and his server is still running fine for nearly a day now. I'll wrap that call in a scheduled task for now when I get back.

**Comment by plasticblues** [comment]
@RoboMWM Well, I started up my test server and... I can't replicate either... I've tried multiple things including but not limited to building a smaller scale of the enderman farm that may have caused my server to crash, no go, tried having a piston go off claim, no go, tried in to claim, no go. Any suggestions?

**Comment by ScuroK** [comment]
@RoboMWM Its me, Dreadonator

Here is the full log:
https://gist.github.com/ScuroK/153c47b69115e8ea6b9c713a121b5e69

**Comment by RoboMWM** [comment]
@plasticblues
> Any suggestions?

No clue either. All I can think of at this point is potentially a plugin conflict. Can you try copying your server and having just GP installed, and see if that farm crashes it?

**Comment by RoboMWM** [comment]
Those of you having issues, try using this build: https://github.com/TechFortress/GriefPrevention/releases/tag/16.5.1

I tried reproducing for an hour, trying various chat things, subdividing claims, resizing claims, making two claims, trying pistons at edges of claims... can't reproduce this issue :/

**Comment by ScuroK** [comment]
did you read my log?

**Comment by RoboMWM** [comment]
Yes I did

**Comment by ScuroK** [comment]
They were talking alot about EntityLWC lock cimmands before the server crashed. 

Maybe its because lwc?

**Comment by RoboMWM** [comment]
@ScuroK not in @MrArcane's case, he doesn't appear to be using that plugin. Anyways, has anyone been having any issues with 16.5.1? So far I have yet to receive a watchdog dump regarding that version.

**Comment by MrArcane** [comment]
@RoboMVM I'm starting to think it's a confliCT aswell, but nothing has changed with my plugins. My test server hasn't been crashing though. Maybe some new code in GP conflicting wit something else? Hm.. I will try to update to 16.5.1 today and see how it goes and let you know

**Comment by MrArcane** [comment]
Sorry about the grammar. On my phone and its horrible

**Comment by sgtcaze** [comment]
@RoboMWM issue appears fixed. Server uptime is at an expected level.

**Comment by RoboMWM** [comment]
Appears to be fixed in [16.5.1](https://github.com/TechFortress/GriefPrevention/releases/tag/16.5.1)
